### DISCUSSION TITLE
Allow top-level ESM usage with Next projects

### DISCUSSION THREAD

**Discussion by JacobLey** [vercel/next.js#32220] [Ideas]
### Describe the feature you'd like to request

NextJS currently requires code be executed as CommonJS.
Packages in `node_modules` do this automagically with custom `package.json` that will declare `"type":"commonjs"` or omit the type (and default to commonjs)

Code in the `.next` directory does not have this luxury, and is stuck using the root packages `type`. And because all code is transpiled to commonjs, that forces to top-level package to also be commonjs.

### Describe the solution you'd like

The best solution to me is simply writing a `package.json` in the `.next` directory with the content `{ "type": "commonjs" }`

That allows NextJS to do everything with CommonJS, while simultaneously freeing up the root package to use whatever type they prefer.

Even if they are already using `commonjs` the new `package.json` is at the least a NOOP

### Describe alternatives you've considered

Change transpiled code to matched the package's type.

This sounds like a lot more work, and probably isn't feasible for use cases like hot reloading, which the current solution requires usage of things like `require.cache`, which do not yet have similar solutions in an ESM environment.

**Comment by JacobLey** [comment]
I currently get around this personally by using a custom `next.config.js` with contents like:

```
'use strict';

const fs = require('fs');
const Path = require('path');

let wrotePackageJson = false;

module.exports = {
    webpack: defaultConfig => {

        if (!wrotePackageJson) {
            fs.mkdirSync(Path.join(opts.dir, '.next'), { recursive: true });
            fs.writeFileSync(Path.join(opts.dir, '.next/package.json'), '{ "type": "commonjs" } ');
            wrotePackageJson = true;
        }
        return defaultConfig;
    },
};
```

Which actually works for now, but overall feels hacky and is depending on timing that is not guaranteed by NextJS

**Reply by kachkaev** [reply]
Related:

- https://github.com/vercel/next.js/issues/24334
- https://github.com/vercel/next.js/discussions/33227

**Reply by kachkaev** [reply]
Potential fix:
- https://github.com/vercel/next.js/pull/33637

**Reply by kachkaev** [reply]
↑ [It seems to work](https://github.com/vercel/next.js/issues/24334#issuecomment-1040748195) 🎉

**Reply by mattxyzeth** [reply]
`opts` is not defined.

**Reply by mattxyzeth** [reply]
oh, it's from the second arg to `webpack`.
